### PR TITLE
Simplify cursor search

### PR DESCRIPTION
**Pull request by aexg**
Store cursor position in engine to reuse it

### CODE REVIEW

File: parser.hpp [hunk 1]
```diff
@@ -47,16 +47,14 @@ constexpr char encode_candy_type(CandyType t)
...
-        if (map_candy_type[i] == c) {
-            return static_cast<CandyType>(i);
-        }
-    }
-
-    if (i == sizeof(map_candy_type)) {
-        throw std::runtime_error("Invalid candy type");
+    switch (c)
+    {
+        case ' ': return CandyType::None;
```
Comment (Jiwan) [inline parser.hpp:52]: Nice, this will avoid a lot of unnecessary iterations.

File: game_engine.hpp [hunk 2]
```diff
@@ -14,11 +15,16 @@ template <std::size_t RowCount, std::size_t ColumnCount>
 class game_engine
 {
 public:
-    CONSTEXPR game_engine(const board<RowCount, ColumnCount>& board, int score, int moves, long long epoch_ms) : 
-        board_(board),
+    CONSTEXPR game_engine(const std::tuple< std::array<std::array<candy, ColumnCount>, RowCount>, int, int, bool, int, int > board_extended, int score, int moves, long long epoch_ms) :
```
Comment (Jiwan) [inline game_engine.hpp]: Although I really like tuples, this might be worth creating a proper `struct`.

File: game_engine.hpp [hunk 3]
```diff
@@ -14,11 +15,16 @@ template <std::size_t RowCount, std::size_t ColumnCount>
...
+        board_(std::get<0>(board_extended)),
         score_(score),
         moves_(moves),
-        rg_(static_cast<std::uint16_t>(epoch_ms))
+        rg_(static_cast<std::uint16_t>(epoch_ms)),
+        hovered_x_(std::get<1>(board_extended)),
+        hovered_y_(std::get<2>(board_extended)),
+        any_selected_(std::get<3>(board_extended)),
```
Comment (Jiwan) [inline game_engine.hpp]: GCC with -Wall might complain that the order of the initializer list is not same as the members.

File: meta_crush_saga.sh [hunk 4]
```diff
@@ -10,7 +10,7 @@ while :
 do
     epoch_ms=$(date +%s%3N)
 
-    g++ -std=c++1z main.cpp -DKEYBOARD_INPUT="$keypressed" -DEPOCH_MS="$epoch_ms" -o renderer
+    g++ -std=c++1z main.cpp -DKEYBOARD_INPUT="$keypressed" -DEPOCH_MS="$epoch_ms" -fconcepts -o renderer
```
Comment (Jiwan) [inline meta_crush_saga.sh:13]: Speaking of -Wall. It might be good to enable here at one point.

**Review by Jiwan — APPROVED** [review]
A lot of good cleanup here! Feel free to merge it as is, or fix the minor issues mentioned.

**Comment by Jiwan** [comment]
I added a reformat_all.sh script on master that use clang-format to streamline the code style.
I will execute it as soon as this PR is merged to avoid any merge conflicts.

File: parser.hpp (hunk 1 above)
Comment (aexg) [reply parser.hpp:52]: Inspired by Elm ;-)

File: game_engine.hpp (hunk 2 above)
Comment (aexg) [reply game_engine.hpp]: Agree! I also thought about using struct, tuple won in choice between rationally good code and cool over-engineered piece. I'll implement struct, later, now at 2:30 A.M. risk for bugs is too high.

File: game_engine.hpp (hunk 3 above)
Comment (aexg) [reply game_engine.hpp]: Turns out it complains about more than just order of initializers! :-) I'll dedicate a branch for warnings clean up.

File: meta_crush_saga.sh (hunk 4 above)
Comment (aexg) [reply meta_crush_saga.sh:13]: Yes, and replace 1z with the official 17 or experimental 2a.

**Comment by aexg** [comment]
I rebased my branch after your commit, all set for merge. I'll wait with other changes to let you clang-format without merging conflict!

File: game_engine.hpp (hunk 2 above)
Comment (aexg) [reply game_engine.hpp]: I made it in time before PR merge! You can now merge version with `struct` instead for `tuple` as you suggested. I can think better at 17:30 than at 2:30 :-)

File: game_engine.hpp (hunk 3 above)
Comment (Jiwan) [reply game_engine.hpp]: Right!

File: game_engine.hpp (hunk 2 above)
Comment (Jiwan) [reply game_engine.hpp]: hehe good job here @aexg !

**Pull request merged**